cycler/__init__.py

```diff
@@ -223,31 +250,36 @@ def _from_iter(cls, label, itr):
         `Cycler`
             New 'base' cycler.
         """
-        ret = cls(None)
+        ret: Cycler[K, V] = cls(None)
         ret._left = list({label: v} for v in itr)
         ret._keys = {label}
         return ret
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: slice) -> Cycler[K, V]:
         # TODO : maybe add numpy style fancy slicing
         if isinstance(key, slice):
             trans = self.by_key()
             return reduce(add, (_cycler(k, v[key]) for k, v in trans.items()))
         else:
             raise ValueError("Can only use slices with Cycler.__getitem__")
 
-    def __iter__(self):
-        if self._right is None:
-            for left in self._left:
-                yield dict(left)
+    def __iter__(self) -> Generator[dict[K, V], None, None]:
+        if self._right is None or self._left is None:
+            if self._left is not None:
+                for left in self._left:
+                    yield dict(left)
```

Comment:
What if `self._left is None` and `self._right is not None`? Can that not happen?

Reply:
I believe that a Cycler with a right child but no left child is invalid, i.e. that all cyclers with only one child (the leaves), it must be the `left` child.

Whether that is enforced or not, is another question.

Certainly, the recommended constructor (`cycler`, the function, rather than `Cycler`, the class) only allows positional args and treats keyword args as label: iterable pairs. As such, that constructor will never yeild a "right only" object

e.g. the `_from_iter` method directly modifies `_left`

Reply:
Note that the previous implementation just directly started iterating on `_left`

Reply:
Actually, I realized that most of the code changes in total could be reverted by simply not _allowing_ `_left` to be None at all. The only way to get the constructor to do it was to do `Cycler(None, None)`. That case was handled in `__init__`, but even if you do it, it fails to even make a `repr`, let alone use _any_ of the methods.

Reply:
Slightly trickier than I thought, as the `left=None` pathway is used, if only temporarily, in `_from_iter`, though that is is the only place where `_left` should be None, and it remedies it right away.

Using empty list (`[]`) as the sentinel for left instead of `None` works, though...